### ISSUE TITLE
twisted/words/test/test_domish.py should use setUp and tearDown instead of setUpClass and tearDownClass

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2741]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2741|
|Type|task|
|Created|2007-07-20 14:30:03Z|

DomishTestListSerializer hooks some domish module globals in a setUpClass and restores the original in tearDownClass.  There's no real reason not to just use setUp and tearDown here.  As long as the code is being changed, it wouldn't hurt to document it and remove the double underscore, too.

<details><summary>Searchable metadata</summary>

```
trac-id__2741 2741
type__task task
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1184941803000000 1184941803000000
changetime__1185709662000000 1185709662000000
version__None None
owner__ 
cc__ralphm cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

That was ridiculous.  The class was entirely pointless.  I deleted it and made the mixin which defines the test methods an actual TestCase.

Code is in setup-teardown-domish-2741

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm commented|
|-|-|

You judge too quickly. This code exists because there used to be more than one serializer in domish. The tests should be the same for all. I left it in there because people are/were looking into writing another serializer.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Was that a review?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

Replying to [ralphm](#note_2):
> You judge too quickly. This code exists because there used to be more than one serializer in domish. The tests should be the same for all. I left it in there because people are/were looking into writing another serializer.

I don't understand the point. ```domish._ListSerializer```, so setting the ```domish.SerializerClass``` to it is useless. The change made by exarkun looks sensible to me.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

Replying to [therve](#note_5):
> Replying to [ralphm](#note_2):
> > You judge too quickly. This code exists because there used to be more than one serializer in domish. The tests should be the same for all. I left it in there because people are/were looking into writing another serializer.
>
> I don't understand the point. ```domish._ListSerializer```, so setting the ```domish.SerializerClass``` to it is useless. The change made by exarkun looks sensible to me.

Grrr: ```domish._ListSerializer``` is already the default, so setting the ```domish.SerializerClass``` to it is useless.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm set owner to @exarkun|
|-|-|

I'll try to explain again. I mostly responded to the _ridiculous_ qualification of the existing tests.

In former times, the were two serializer classes, `_Serializer` and `_ListSerializer` with the former being the default. The difference was the way the serialization was constructed (`StringIO.StringIO.write` vs. `list.append` with a final `''.join(list)`. The unit tests had to be identical for both so I that's the reason `SerializerTests` mixin was created.

Then, in [13642], the `_Serializer` class was removed because `_ListSerializer` turned out to be much faster and it was more work to keep the two serializers in sync. I left the unit test framework as it is, because I could imagine other serializers to pop up in the future.

So yeah, currently there is only one serializer which is naturally also the default.

Recently, though, the group of people at Chesspark (user `jack`) worked on optimizations in domish. One exercise was the use of `libxml2` to do the serialization, and while I haven't see their code, I'm sure they could easily reuse the existing test framework by just adding a new test case for their serializer.

With questions of using some existing XML library instead of domish every few months, I can imagine a second serializer to be added to domish at some point in the future.

So, the only thing I was saying is that the existing stuff is there for a reason that is not ridiculous. That said, I'd be ok with using `setUp` and `tearDown` instead of `classSetUp` and `classTearDown` if that makes people more happy.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @ralphm|
|-|-|

Thanks for clarifying the history of the code in question.  Hope no offense was taken by my original comment, as none was intended. :)  I'm quite familiar with circumstances where refactoring leaves code in a superficially odd shape.

I don't think the changes in the branch really drop a significant amount of utility from the factoring of the test suite.  Actually, I think the tests should be factored such that there's no need to set this module global in order to test different serializers.  So I'd still like to merge the branch as-is.  SerializerClass is only used in one place in domish.py - Element.toXml, so it should be easy to parameterize on a per-instance (or maybe even per-call) basis, rather than playing with the global attribute.  Does this sound like a reasonable plan, once someone contributes another serializer (I think this also makes it easier for people to work on their own serializers externally, since it removes the global mutable state "API" for providing a new one)?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm set owner to @exarkun|
|-|-|

No offense taken. You have to try harder for that ;-)

Your rational makes sense to me. The branch looks ok, so please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [20867]) Merge setup-teardown-domish-2741

Author: exarkun
Reviewer: therve, ralphm
Fixes [#2741](https://github.com/twisted/twisted/issues/2741)

Remove DomishTestListSerializer and its test-defining mixin
into a TestCase so its test methods will be run.  Drop the
setUpClass and tearDownClass methods, since they don't do
anything as the code currently stands.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
